### PR TITLE
DBZ-4783: Support for multiple databases and tasks in the SQL Server connector

### PR DESCRIPTION
**Pull request by morozov**
Change summary:
1. Manage change tables for each partition individually.
2. Expose the actual task id to the metrics in order to be able to run multiple tasks.
3. Run the connector with multiple databases and multiple tasks.

Note, running multiple tasks is _not_ necessary for capturing multiple databases. Furthermore, running multiple tasks cannot be tested in the current test suite. The only reason why it's contributed here is that extracting the support for multiple tasks in a separate patch would require extra work (it was originally implemented like this). I can remove it from the patch, if necessary.

**TODO**:
- [x] Provide an integration test for the multi-partition scenario.

### Manual testing of the multi-database and multi-task configuration

1. Modify `docker-compose-sqlserver.yaml` from the [tutorial](https://github.com/debezium/debezium-examples/tree/main/tutorial#using-sql-server) to use the Docker image built from this PR.
2. Start the services:
   ```shell
   export DEBEZIUM_VERSION=1.9
   docker-compose -f docker-compose-sqlserver.yaml up
   ```
3. Initialize test databases:
   ```shell
   docker-compose -f docker-compose-sqlserver.yaml exec -T sqlserver bash -c '/opt/mssql-tools/bin/sqlcmd -U sa -P $SA_PASSWORD' << 'EOF'
   CREATE DATABASE testDB1;
   GO
   USE testDB1;
   EXEC sys.sp_cdc_enable_db;
   
   CREATE TABLE products (
     id INTEGER IDENTITY(101,1) NOT NULL PRIMARY KEY,
     name VARCHAR(255) NOT NULL,
     description VARCHAR(512),
     weight FLOAT
   );
   INSERT INTO products(name,description,weight)
     VALUES ('scooter','Small 2-wheel scooter',3.14);
   
   EXEC sys.sp_cdc_enable_table @source_schema = 'dbo', @source_name = 'products', @role_name = NULL, @supports_net_changes = 0;
   GO
   
   CREATE DATABASE testDB2;
   GO
   USE testDB2;
   EXEC sys.sp_cdc_enable_db;
   
   CREATE TABLE customers (
     id INTEGER IDENTITY(1001,1) NOT NULL PRIMARY KEY,
     first_name VARCHAR(255) NOT NULL,
     last_name VARCHAR(255) NOT NULL,
     email VARCHAR(255) NOT NULL UNIQUE
   );
   INSERT INTO customers(first_name,last_name,email)
     VALUES ('Sally','Thomas','sally.thomas@acme.com');
   
   EXEC sys.sp_cdc_enable_table @source_schema = 'dbo', @source_name = 'customers', @role_name = NULL, @supports_net_changes = 0;
   GO
   EOF
   ```

4. Start the connector:

   ```shell
   curl -i -X POST -H "Accept:application/json" -H "Content-Type:application/json" http://localhost:8083/connectors/ -d@- << 'EOF'
   {
       "name": "inventory-connector",
       "config": {
           "connector.class" : "io.debezium.connector.sqlserver.SqlServerConnector",
           "tasks.max" : "2",
           "database.server.name" : "server1",
           "database.hostname" : "sqlserver",
           "database.port" : "1433",
           "database.user" : "sa",
           "database.password" : "Password!",
           "database.names" : "testDB1,testDB2",
           "database.history.kafka.bootstrap.servers" : "kafka:9092",
           "database.history.kafka.topic": "schema-changes.inventory"
       }
   }
   EOF
   ```

5. Check that the connector and both tasks are running:

   ```shell
   curl -s http://localhost:8083/connectors/inventory-connector/status
   ```

6. Check that each task exposes metrics for its subset of databases:
   Multi-partition and multi-task metrics

7. Run a consumer and confirm that both databases were snapshotted:

   ```shell
   docker-compose -f docker-compose-sqlserver.yaml exec kafka /kafka/bin/kafka-console-consumer.sh \
     --bootstrap-server kafka:9092 \
     --from-beginning \
     --property print.key=true \
     --whitelist 'server1\.(testDB\d+)\..*'
   ```

8. Make changes in both databases and confirm that the changes from both databases are captured:

   ```shell
   docker-compose -f docker-compose-sqlserver.yaml exec -T sqlserver bash -c '/opt/mssql-tools/bin/sqlcmd -U sa -P $SA_PASSWORD' << 'EOF'
   UPDATE testDB1.dbo.products SET weight = 3.15 WHERE id = 101;
   UPDATE testDB2.dbo.customers SET first_name = 'Molly' WHERE id = 1001;
   EOF
   ```

### CODE REVIEW

**Comment by gunnarmorling** [comment]
Thanks a lot, @morozov! Will take a loop asap.

@jpechane, as per Sergei's mail, could you suggest a few key tests which should be adapted (well, copied, I suppose) to ensure the multi-partition mode works?

File: debezium-connector-sqlserver/src/main/java/io/debezium/connector/sqlserver/SqlServerConnector.java [hunk 1]
```diff
@@ -90,8 +128,7 @@ protected void validateConnection(Map<String, ConfigValue> configValues, Configu
         final SqlServerConnectorConfig sqlServerConfig = new SqlServerConnectorConfig(config);
 
         if (Strings.isNullOrEmpty(sqlServerConfig.getDatabaseName())) {
-            throw new IllegalArgumentException("Either '" + SqlServerConnectorConfig.DATABASE_NAME
-                    + "' or '" + SqlServerConnectorConfig.DATABASE_NAMES
+            throw new IllegalArgumentException("Either '" + DATABASE_NAME + "' or '" + DATABASE_NAMES
```
Comment (gunnarmorling) [inline debezium-connector-sqlserver/src/main/java/io/debezium/connector/sqlserver/SqlServerConnector.java:131]: Rather a configuration error should be added (see connection validation failure in this method below).

File: debezium-connector-sqlserver/src/main/java/io/debezium/connector/sqlserver/SqlServerPartition.java [hunk 2]
```diff
@@ -70,17 +77,30 @@ public String toString() {
...
-            return Collections.singleton(new SqlServerPartition(
-                    connectorConfig.getLogicalName(),
-                    connectorConfig.getDatabaseName(),
-                    connectorConfig.isMultiPartitionModeEnabled()));
+            String serverName = connectorConfig.getLogicalName();
+            boolean multiPartitionMode = connectorConfig.isMultiPartitionModeEnabled();
+
+            List<String> databaseNames;
+
+            if (multiPartitionMode) {
```
Comment (gunnarmorling) [inline debezium-connector-sqlserver/src/main/java/io/debezium/connector/sqlserver/SqlServerPartition.java:94]: See above.

**Review by gunnarmorling — COMMENTED** [review]
Thanks a lot, @morozov! A few stylistic comments inline. Still needs to do some testing using Compose (thanks for sharing the instructions for that!).

File: debezium-connector-sqlserver/src/main/java/io/debezium/connector/sqlserver/SqlServerConnector.java [hunk 3]
```diff
@@ -51,29 +57,61 @@ public void start(Map<String, String> props) {
 
     @Override
     public List<Map<String, String>> taskConfigs(int maxTasks) {
-        if (maxTasks > 1) {
-            throw new IllegalArgumentException("Only a single connector task may be started");
+        if (maxTasks > 1 && !properties.containsKey(DATABASE_NAMES.name())) {
+            throw new IllegalArgumentException("Only a single connector task may be started in single-partition mode");
         }
```
[the remark this answers is not in the document]
Comment (morozov) [reply debezium-connector-sqlserver/src/main/java/io/debezium/connector/sqlserver/SqlServerConnector.java:62]: `int maxTasks` is unavailable in the scope of `validateAllFields()`. Making it available there would require replication of the Kafka Connect logic (sourcing it from a certain configuration parameter, considering the default value, etc). I believe it makes more sense to keep this logic here and remove it entirely in 2.0.

File: debezium-connector-sqlserver/src/main/java/io/debezium/connector/sqlserver/SqlServerConnector.java (hunk 1 above)
Comment (morozov) [reply debezium-connector-sqlserver/src/main/java/io/debezium/connector/sqlserver/SqlServerConnector.java:131]: There is actually already validation of both parameters:

1. For `DATABASE_NAME`:
https://github.com/debezium/debezium/blob/2d14b5b09f4af68f6070fbfc4e4a62cd323290f6/debezium-connector-sqlserver/src/main/java/io/debezium/connector/sqlserver/SqlServerConnectorConfig.java#L235
2. And for `DATABASE_NAMES`: https://github.com/debezium/debezium/blob/2d14b5b09f4af68f6070fbfc4e4a62cd323290f6/debezium-connector-sqlserver/src/main/java/io/debezium/connector/sqlserver/SqlServerConnectorConfig.java#L243

So we just need to early return if any of these fields is invalid as it was done prior to https://github.com/debezium/debezium/pull/2604: https://github.com/debezium/debezium/blob/ade15cd8f3c76eb8cbc0c5240716dd92d6c6821b/debezium-connector-sqlserver/src/main/java/io/debezium/connector/sqlserver/SqlServerConnector.java#L82-L85

**Comment by morozov** [comment]
@gunnarmorling I addressed your review comments and added a basic snapshotting/streaming test. Please take a look.

File: debezium-connector-sqlserver/src/main/java/io/debezium/connector/sqlserver/SqlServerConnector.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (gunnarmorling) [reply debezium-connector-sqlserver/src/main/java/io/debezium/connector/sqlserver/SqlServerConnector.java:60]: Ok, makes sense.

File: debezium-connector-sqlserver/src/main/java/io/debezium/connector/sqlserver/SqlServerConnector.java [hunk 4]
```diff
@@ -51,29 +57,61 @@ public void start(Map<String, String> props) {
...
-            final String realDatabaseName = connection.retrieveRealDatabaseName(databaseName);
-            if (!sqlServerConfig.isMultiPartitionModeEnabled()) {
-                taskConfig.put(SqlServerConnectorConfig.DATABASE_NAME.name(), realDatabaseName);
+            if (multiPartitionMode) {
+                databaseNames = Arrays.asList(properties.get(DATABASE_NAMES.name()).split(","));
             }
             else {
-                taskConfig.put(SqlServerConnectorConfig.DATABASE_NAMES.name(), realDatabaseName);
+                databaseNames = Collections.singletonList(properties.get(DATABASE_NAME.name()));
             }
```
[the remark this answers is not in the document]
Comment (gunnarmorling) [reply debezium-connector-sqlserver/src/main/java/io/debezium/connector/sqlserver/SqlServerConnector.java]: Ok, cool.

**Review by gunnarmorling — APPROVED** [review]
LGTM, merging.

@morozov, is there anything else missing before we can declare multi-partitioning mode as functional? I.e. could we announce it as a new feature in the Beta1 release with this PR in?

**Comment by morozov** [comment]
There are a couple of minor things left:

1. Add task id and partition to logging context. Originally implemented in https://github.com/sugarcrm/debezium/pull/80, the current state is https://github.com/sugarcrm/debezium/commit/e3774be18adea7af33f592136c98f5dc0d0952eb. It may need some minor tweaking to make it work with https://github.com/debezium/debezium/pull/3103.

2. The partition-scoped `CapturedTables` metric will likely expose all tables captured by the task, not only the ones that belong to the corresponding partition:
   https://github.com/debezium/debezium/blob/2d14b5b09f4af68f6070fbfc4e4a62cd323290f6/debezium-connector-sqlserver/src/main/java/io/debezium/connector/sqlserver/metrics/SqlServerStreamingPartitionMetrics.java#L26-L29 https://github.com/debezium/debezium/blob/2d14b5b09f4af68f6070fbfc4e4a62cd323290f6/debezium-core/src/main/java/io/debezium/pipeline/meters/StreamingMeter.java#L43-L46 We haven't attempted to address or even reproduce this issue since we're not using this metric. Some of our MySQL connectors capture a couple of hundred thousand tables.

And this is it, there are no more changes related to the multi-partition mode in our fork. I will file Jira cases for each of them but I believe it shouldn't block the announcement.

**Comment by gunnarmorling** [comment]
Thx a lot for logging these follow-up, issues, @morozov! Also a big thank you for that awesome PR description, it just came in super-handy for creating a screenshot of the metrics in JMC.

**Pull request merged**